Add hashing function, SeededRng, and improve naming [MAID-2914, MAID-2915, MAID-2916]

nbaksalyar opened the pull request.

nbaksalyar committed:
refactor/rename: rename structures and functions

Use better names reflecting the intended use of the structs and functions.
Use shorter feature name (mock instead of use-mock-crypto).

MAID-2916

nbaksalyar committed:
feat/hash: add hashing function to the public API

Also provide the mock version.

MAID-2914

nbaksalyar committed:
feat/seeded-rng: add the fast pseudo-random generator for mock crypto

MAID-2915

Fraser999 commented on src/seeded_rng.rs:
```diff
@@ -0,0 +1,229 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// http://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
```
https ?

Fraser999 commented on .travis.yml:
```diff
@@ -27,7 +27,7 @@ script:
         set -x;
         cargo test --verbose --release &&
-        cargo test --verbose --release --features use-mock-crypto
+        cargo test --verbose --release --features mock
```
I think we can add a line similar to https://github.com/Fraser999/maidsafe_utilities/blob/master/.travis.yml#L34 to test the seeded rng in a standalone test run.

Fraser999 commented on appveyor.yml:
```diff
@@ -32,4 +32,4 @@ build_script:
 test_script:
   - cargo test --verbose --release
-  - cargo test --verbose --release --features use-mock-crypto
+  - cargo test --verbose --release --features mock
```
Similar to the Travis script, we can add a line like https://github.com/Fraser999/maidsafe_utilities/blob/master/appveyor.yml#L35 for the seeded rng.

Fraser999 commented on src/lib.rs:
```diff
@@ -176,26 +198,26 @@ impl PublicId {
 
     /// Returns the public part of the secret key set.
-    pub fn public_id(&self) -> &PublicId {
+    pub fn public_id(&self) -> &PublicKeys {
```
I suppose this should be `public_keys()` now, or just `public()`?

Fraser999 commented on src/lib.rs:
```diff
@@ -33,30 +33,46 @@
 use std::sync::Arc;
 
+/// Represents public signature key.
+pub type PublicSignKey = [u8; 32];
```
Did we agree to export something like `const PUBLIC_SIGN_KEY_BYTES: usize = 32;` and use that here?

ustulation replied:
I didn't write it in my notes as getting agreed, rather as point was raised but it's Routing that wants to set this limit and not the other way round - so if it was configurable then yes it's helpful but if not, the constant is basically Routing's and this is just here as a (in absence of a better way) a reverse dependency. Maybe instead of the constant we could wait for future versions of Rust to make constants configurable by user-libs and then make this change. It's not a big point so I'm OK if we want a constant just now too.

nbaksalyar committed:
refactor/all: address review comments

1. Add CI checks for SeededRng
2. Fix non-secure links
3. Rename SecretKeys::public_id to SecretKeys::public_keys
